### ISSUE TITLE
referral program

### ISSUE THREAD

**Issue by jaketoolson** [kabooodle/foundation#33]
The referral program is the main source of (self) marketing. Similar to Dropbox, users can refer friends to Kabooodle and earn up $25 in credits.  For each person they refer, both the new referral and the user earn $5 in credits each!  We will likely make lots of changes annually to this.

In order for users to be able to refer friends, we need to create a URL for them to pass around.  This URL should have their username in it so that when the URL is hit, we can grab the username etc;  With Laravel, this can be done using route middleware.  There is already a referral middleware that I setup for testing purposes that can be reused/modified (Http\Middleware\ReferralProgramMiddleware.php). 

Basically, when a user hits the referral URL, the middleware intercepts the request and parses the username from it, storing it into the session.  When the user registers, only then do we check for this referral shit and do logic with it.
